### PR TITLE
Feature/issue 02 models and dependency injectors

### PR DESCRIPTION
**Pull request by olpie101**
Create model for db on phone as well as REST services and dependency injectors to make app management easier. This brach is a set up branch. So don't think it will require much review.

Decided to write the app using `kotlin`, an open-source JVM language that is 100% interoperable with `Java`. The runtime is very small so it won't add a lot o weight to the apk.

For an introduction to `kotlin` look [here](http://www.kotlinlang.org) and the docs [here](http://kotlinlang.org/docs/reference/), however to speed people up a little here are links to common sections in the docs:
- [Basic types](http://kotlinlang.org/docs/reference/basic-types.html)
- [Packages](http://kotlinlang.org/docs/reference/packages.html) - for `import` aliasing
- [Control Flow](http://kotlinlang.org/docs/reference/control-flow.html) - for special control syntax such as the `when` statements
- [Classes](http://kotlinlang.org/docs/reference/classes.html) - for classes and inheritance syntax
- [Properties and Fields](http://kotlinlang.org/docs/reference/properties.html) - for differences between `val` and `var` and how getter and setters are implemented
- [Extensions](http://kotlinlang.org/docs/reference/extensions.html) - explains how function extensions work
- [Functions](http://kotlinlang.org/docs/reference/functions.html) - for function syntax
- [Lambdas](http://kotlinlang.org/docs/reference/lambdas.html)

Any issues feel free to ask.

### CODE REVIEW

**Comment by olpie101** [comment]
Just finished creating models

**Comment by JayH5** [comment]
Generally, you don't need to include the `.idea` directory or any `*.iml` files in git when using IntelliJ/AS. There is of quite a lot of churn in those files and they're usually not needed by people to import the project.

**Comment by olpie101** [comment]
True. AS decided that it would only add 2 files to the `.gitignore` file from the `.idea` directory will change it.

**Comment by olpie101** [comment]
Set up complete. Tests to follow. Large number of insertions but a large portion are the java classes that represent the DB models which some have large number of fields and all need getters and setters.

NB: Due to current issues with `kotlin` and annotation processors, certain classes have to be written in `Java` ,:cry: :sob: :disappointed:, such as the model classes created in [this commit](https://github.com/praekelt/mama-android/commit/466fad74f5a1f8ac8cba216ffaec4e3471dc27c9), everything else is written in `kotlin` :smile: :smiley: :grin: :grinning:.

**Comment by olpie101** [comment]
All test pass :smile: :grinning: . _On a side note, don't try and get tests to work on Android they are a serious pain. If/When Praekelt gets into the Android space docs should be written on testing_

NB: Once again due to issues with the `gradle` build system not liking `kotlin` The tests need to be written in `Java`

**Comment by olpie101** [comment]
@smn fully ready for review. Though not quite sure who to ping. I guess you can ping the appropriate people for me :smile:, though not quite sure if people want to look through ~ over 1500 lines of code :scream: :fearful:.

File: app/src/test/java/za/foundation/praekelt/mama/api/rest/UCDServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        FlowManager.destroy();
+    }
+
+    @Test
+    public void testGetRepoStatus() {
+        long currentTime = System.currentTimeMillis();
+
+        Observable.just(ucdService.getRepoStatus().toBlocking().single())
+                .subscribe(status -> {
+                    assertEquals("repo name", Constants.REPO_NAME, status.getName());
```
Comment (JayH5) [inline app/src/test/java/za/foundation/praekelt/mama/api/rest/UCDServiceTest.java]: Are you using `assertj` or not?

``` java
assertThat(status.getName()).isEqualTo(Constants.REPO_NAME);
```

Comment (olpie101) [inline app/src/test/java/za/foundation/praekelt/mama/api/rest/UCDServiceTest.java]: Thanks. Switched it to assertj

File: app/src/test/java/za/foundation/praekelt/mama/api/rest/UCDServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                });
+    }
+
+    @Test
+    public void testCloneRepo() {
+        Observable.just(ucdService.cloneRepo().toBlocking().single())
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(repo -> {
+                    assertThat(repo.getCommit()).isNotEqualTo("");
```
Comment (JayH5) [inline app/src/test/java/za/foundation/praekelt/mama/api/rest/UCDServiceTest.java:70]: Ok, but now you can _use_ assertj...

``` java
assertThat(repo.getCommit()).isNotEmpty();
assertThat(repo.getCategories()).isNotEmpty().hasSize(3);
```

...assuming `getCategories()` returns an `Iterable`...

Comment (olpie101) [inline app/src/test/java/za/foundation/praekelt/mama/api/rest/UCDServiceTest.java:70]: thanks changed

**Comment by smn** [comment]
My only remark at this point is to mock out network calls.

**Comment by olpie101** [comment]
@smn mocked network calls. However the travis script doesn't seem to fire when I push

**Pull request merged**